### DISCUSSION TITLE
Extensions in other languages

### DISCUSSION THREAD

**Discussion by diogox** [Ulauncher/Ulauncher#895] [Ideas]
I've vocalized this in another issue as a sort of off-topic, but figured it would be best to make it its own issue. 

**The idea**
Changing the architecture so that virtually any language could be used to make extensions, provided that someone creates a lib to communicate via websockets with the server. 

**How it could be done**
As far as I could tell, and I'm no python programmer (part of the reason I want this "feature"), all it would take is trading pickle for JSON serialization/deserialization. And changing how the server starts the extensions so that it is not exclusive to python. Possibly also the github download so that it compiles the code in case of compiles languages (that part could be specified in the manifest). 

I think this would really open a lot of possibilities and, in some cases, make the extensions faster. I'm available to make the Rust lib for this if it gets implemented.

**Comment by friday** [comment]
I think this would be a nice addition if it doesn't increase the maintenance or documentation burden. Loading binaries or [shared objects](https://doc.rust-lang.org/1.5.0/book/rust-inside-other-languages.html#python) for example means you could target all compiled languages. Extension developers would have to build releases though. If not, then Ulauncher would need to depend on very large build dependencies. Rust alone would be ~250mb of extra dependencies. Go, Haskell and Ocaml needs something similar. Installing extensions would be a slower since they also have to build. This is more negative than positive for end users imo.

The changes needed to support this would also solve #285 I think. And extensions in compiled languages would likely have further direct positive effect on resource usage too.

Edit: Rephrased and removed repetition and redundancies.

**Comment by gornostal** [comment]
That's a nice feature to have.
Before I look into that, I need to address a few more important things, like extension API stability, Python 3 support, file and network permissions for extensions.

**Comment by Jaakkonen** [comment]
This would also possibly make it easier to use  `mypyc` and `cython` for compiling the extensions to Python C extensions which allows them to run faster and with a smaller memory footprint.
